### PR TITLE
 [Distributed] [7/N] Fix clang-tidy warnings in torch/csrc/distributed/c10d 

### PR DESCRIPTION
**Pull request by cyyever**
This PR continues to clean clang-tidy warnings in torch/csrc/distributed/c10d, following #124701. 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @malfet @seemethere

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroup.hpp [hunk 1]
```diff
@@ -59,10 +59,11 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
     std::chrono::milliseconds timeout;
 
     // backend name
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-const-or-ref-data-members)
     const std::string backend;
   };
 
-  enum BackendType {
+  enum BackendType : uint8_t {
```
Comment (malfet) [inline torch/csrc/distributed/c10d/ProcessGroup.hpp:66]: Hmm, what clang-tidy rule is that?

**Review by malfet — APPROVED** [review]
Mostly LGTM, but before merging please add some explanation why `fmt` dependency must be added to torch_cuda/torch_hip (is that really the first use of `fmt`. And perhaps one can move this change into a separate PR from the rest of the static casts)

Also, what clang-tidy rules mandates typed `enum class`es?

File: torch/csrc/distributed/c10d/ProcessGroup.hpp (hunk 1 above)
Comment (cyyever) [reply torch/csrc/distributed/c10d/ProcessGroup.hpp:66]: The newer clang-tidy 18 has added performance-enum-size to warn enum size, see
https://clang.llvm.org/extra/clang-tidy/checks/performance/enum-size.html

**Comment by danthe3rd** [comment]
Hi,
This is breaking 3rd C++ libraries which include `<torch/csrc/distributed/c10d/ProcessGroup.hpp>`, like in xFormers:
https://github.com/facebookresearch/xformers/blob/fad50d49834ab18dd137acc727bd4d567ff17842/xformers/csrc/boxing_unboxing.cpp#L8
Failure:
```
    In file included from /ghrunner/shared/tmp/8878728050/lib/python3.11/site-packages/torch/include/torch/csrc/distributed/c10d/Backend.hpp:11,
                     from /ghrunner/shared/tmp/8878728050/lib/python3.11/site-packages/torch/include/torch/csrc/distributed/c10d/ProcessGroup.hpp:3,
                     from /ghrunner/_work/xformers/xformers/xformers/csrc/boxing_unboxing.cpp:8:
    /ghrunner/shared/tmp/8878728050/lib/python3.11/site-packages/torch/include/torch/csrc/distributed/c10d/Utils.hpp:7:10: fatal error: fmt/format.h: No such file or directory
        7 | #include <fmt/format.h>
          |          ^~~~~~~~~~~~~~
    compilation terminated.
```

cc @malfet @ezyang

File: torch/csrc/distributed/c10d/Utils.hpp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #include <c10/util/Exception.h>
 #include <c10/util/accumulate.h>
 #include <c10/util/irange.h>
+#include <fmt/format.h>
```
Comment (xwang233) [inline torch/csrc/distributed/c10d/Utils.hpp:7]: This change added the FMT header, but it's not included in torch nightly wheels or bundled with `python setup.py install`, and caused pytorch extension build to fail.

Comment (xwang233) [reply torch/csrc/distributed/c10d/Utils.hpp:7]: cc @malfet @atalman

**Comment by cyyever** [comment]
Using linfmt involves installing it in the build systems, and it's not a good time to change them at the present.

**Pull request closed without merging**